File: adapters/base_adapter.py
"""
Base Adapter

Foundation class for all service adapters.
"""

from typing import Dict, Any
from django.conf import settings
import logging
import time

logger = logging.getLogger(__name__)


class BaseAdapter:
    """
    Base class for service adapters.

    Provides common functionality:
    - Feature flag checking
    - Metrics tracking
    - Error handling patterns
    """

    def __init__(self, service_name: str):
        self.service_name = service_name
        self.use_microservice = self._get_feature_flag()
        self.fallback_enabled = self._get_fallback_flag()

    def _get_feature_flag(self) -> bool:
        """Get feature flag for this service."""
        flag_name = f'USE_{self.service_name.upper()}_MICROSERVICE'
        return getattr(settings, flag_name, False)

    def _get_fallback_flag(self) -> bool:
        """Get fallback flag for this service."""
        flag_name = f'{self.service_name.upper()}_MICROSERVICE_FALLBACK'
        return getattr(settings, flag_name, True)

    def _track_request(
        self,
        operation: str,
        provider: str,
        success: bool,
        duration: float,
        **metadata
    ):
        """
        Track adapter request metrics.

        Args:
            operation: Operation name (e.g., 'send_sms')
            provider: 'microservice' or 'legacy'
            success: Whether operation succeeded
            duration: Duration in seconds
            **metadata: Additional metadata
        """
        try:
            from core.metrics import track_metric

            track_metric('adapter.request', {
                'service': self.service_name,
                'operation': operation,
                'provider': provider,
                'success': success,
                'duration_ms': duration * 1000,
                **metadata
            })
        except Exception as e:
            logger.warning(f"Failed to track metrics: {e}")

    def _execute_with_tracking(
        self,
        operation: str,
        func,
        *args,
        **kwargs
    ) -> Dict[str, Any]:
        """
        Execute function with automatic tracking.

        Args:
            operation: Operation name
            func: Function to execute
            *args, **kwargs: Function arguments

        Returns:
            Function result with 'provider' key added
        """
        start_time = time.time()
        result = func(*args, **kwargs)
        duration = time.time() - start_time

        self._track_request(
            operation=operation,
            provider=result.get('provider', 'unknown'),
            success=result.get('success', False),
            duration=duration
        )

        return result

File: smsly-rust/smsly-services/src/adapters/base_adapter.rs
use crate::config::Settings;
use serde_json::Value;
use smsly_core::metrics::track_metric;
use std::collections::HashMap;

pub struct BaseAdapter {
    pub service_name: String,
    pub settings: Settings,
    pub use_microservice: bool,
    pub fallback_enabled: bool,
}

impl BaseAdapter {
    pub fn new(service_name: String, settings: Settings) -> Self {
        let use_microservice = settings.is_microservice_enabled(&service_name);
        let fallback_enabled = settings.is_fallback_enabled(&service_name);

        Self {
            service_name,
            settings,
            use_microservice,
            fallback_enabled,
        }
    }

    pub fn track_request(
        &self,
        operation: &str,
        provider: &str,
        success: bool,
        duration: f64,
        metadata: Option<HashMap<String, Value>>,
    ) {
        let mut meta = metadata.unwrap_or_default();
        meta.insert(
            "service".to_string(),
            Value::String(self.service_name.clone()),
        );
        meta.insert(
            "operation".to_string(),
            Value::String(operation.to_string()),
        );
        meta.insert("provider".to_string(), Value::String(provider.to_string()));
        meta.insert("success".to_string(), Value::Bool(success));
        if let Some(n) = serde_json::Number::from_f64(duration * 1000.0) {
            meta.insert("duration_ms".to_string(), Value::Number(n));
        }

        track_metric("adapter.request", meta);
    }
}
